### ISSUE TITLE
Proposed improvement to new work item experience

### ISSUE THREAD

**Issue by qodfathr** [openshiftio/openshift.io#1225]
There have been on-going discussions about entering new work items on the backlog. The "quick add" method permits the quick creation of WIs with few details, and is considered nice for someone like a Product Manager. A more complete "add via the full details form" is seen as better for people like developers, and is more closely aligned with the experience in services such as GitHub.

I propose a change for consideration which may address both.

*Quick Add Change*
1. Remove the behavior on Quick Add that it initially starts in a "+ Create Work Item" state, requiring the user to click it to reveal the quick add UI
1. Change the quick add fields so that it only takes a work item type and a title (i.e. remove the ability to add the description)
1. Get rid of the Green Checkmark and the Red X buttons as part of the quick add
1. Replace those two buttons with textual buttons -- `Add` and `Add and Open`
1. When focus is on Title, `[Enter]` behaves as `Add` and `[SHIFT+ENTER]` or `[CTRL+ENTER]` behaves as `Add and Open`

There is really no need for the red X -- "canceling" is implied by not clicking one of the `Add` buttons (or typing some form of `[Enter]` when focus is in the Title field)
The behavior of `Add` is to add the WI to the backlog.
The behavior of `Add and Open` is to add the WI to the backlog and then immediately open the WI details screen.

*Work Item Details Screen Change*
1. Change behavior of form so that it is not auto-saving with every field change.
1. Change behavior of X close button to behave same as Cancel button as described below.
1. Offer a new set of buttons (which could just be a single button that has a drop-down providing the choices):
   1. `Save and Close` -- save the work item and close the detail form, returning to the previous view. *Disabled* if there have not been any changes made to the work item (e.g. it is not dirty).
   1. `Save` -- save all current changes to the work item. *Disabled* is work item is not dirty. Detail form remains open, and, once saved, is no longer dirty.
   1. `Save and New` -- save all current changes to the work item, and open a new details form for a new work item. *Disabled* if work item is not dirty.
   1. Cancel -- Discard all changes (if any) and return to the previous view. *Warn* user if work item is dirty and confirm that user wants changes to be lost.

Under this proposal, the PM persona would stay in the Quick Add world, typing new titles for WIs and clicking Add (or just pressing Enter). The Dev persona would type the first title into Quick Add, then either click `Add and Open` or type `[SHIFT+ENTER]`, revealing the full detail form. From here, the Dev persona can continue to use the `Add and New` button to create additional WIs, each with full details.

Initial comments appreciated from @catrobson / @sdash-redhat and other members of UXD team.

**Comment by ldimaggi** [comment]
This souds good! One question - for the quick add path - will we provide* the user with a default work item type? 

* Or, allow the user to configure this choice?

**Comment by qodfathr** [comment]
@ldimaggi the Quick Add would still require the selection of the Work Item Type, as it is today.

**Comment by sdash-redhat** [comment]
+1
This sounds a good idea..best from the both worlds of personae. 
"Add and Open" is a great entry point for additional info entry...  May be a label on this new button like  "Add More Info" can hint that the user can add more info ... 

In work details screen ..I wonder if we can reduce the number of buttons (currently 4 ones each one having a "save" prefix .. or may be we can think of changing the labels that make it lear to the user-obvious .. or may reduce the numbers. Any thoughts on this?

Overall this is a great idea.

**Comment by qodfathr** [comment]
@sdash-redhat, I agree that 4 buttons is a bit crazy. :) Does UXD have a recommendation? For example, what are UXD's thoughts on a combo-control button which includes a drop-down, thereby 'stacking' several button operations on a single button?

And if that's acceptable, would the choice become sticky? i.e., if the default is `Save and Close`, but I change it do be `Save and New`, would it remain `Save and New` on the newly opened work item? How about if I leave this screen and come back later? Does that revert back to `Save and Close` or is my change to `Save and New` permanent (until I change it again)?

**Comment by sdash-redhat** [comment]
Yeah...  sticky option also sounds more close... Let me draw these and play-around ..that might light some bulb ;) I will get back tomorrow ...

**Comment by sdash-redhat** [comment]
Here is quick-preview of how they might look like ...

![image](https://user-images.githubusercontent.com/27761023/32145911-f43f065c-bcf5-11e7-8331-41504658b6a7.png)
If the user clicks on "Save & Open" he lands n the detailed view 

The following detailed view shows all the options of save as a split button (PF pattern)
![image](https://user-images.githubusercontent.com/27761023/32145923-345ba11e-bcf6-11e7-9d52-ec155b5702ee.png)
![image](https://user-images.githubusercontent.com/27761023/32145952-9cc43c52-bcf6-11e7-9cd4-fd9e97789a12.png)


If we are to show one button and other as options of a drop down that might look like the following ...

![image](https://user-images.githubusercontent.com/27761023/32145947-8a36881a-bcf6-11e7-9518-745794042421.png)
![image](https://user-images.githubusercontent.com/27761023/32145971-c603076a-bcf6-11e7-9cc0-86c4bb9f5990.png)



 I am thinking if "Save and New" is really needed upfront in detailed page..as by using "Save & Close" he ends up in the list page, where he can start creating another  workitem without any further button clicks (remember now quick add upfront provides text field to enter)

So should we remove "Save & New" from the button list ..that will also make it easier even to use a split button.

**Comment by sdash-redhat** [comment]
Some minor tweak in labels --  may be we can use "Save" and "Add Details"  (Instead of "Add" . & "Add & Open" ) which will not sound monotonous and give similar meaning. 
Also to reset the text box, will use an 'X' icon in the text box when the text is filled in. Rest everything remains same.

**Comment by qodfathr** [comment]
@sdash-redhat re: "Add & Open" vs. "Add Details".

`Add & Open` makes it reasonably clear that the item is being *added* and then being *opened*. When the user reaches the details page and clicks the Cancel button, it should be self-evident that the item was already added, so clicking `Cancel` does *not* result in no item being added to the backlog. It is simply canceling any changes made to the details.

`Add Details` could be interpreted as "I've specified the work item type and the title, and now I want to add all of the other details." My concern is that it is not self-evident that the item *has also* been saved as a result of clicking the button. A user might rightfully expect that, upon visiting the details page, clicking `Cancel` will result in no item having been added to the backlog.

I may be over-thinking it, but that would be my only concern. The details button is doing two separate actions -- it is adding a new item to the backlog and also opening the details form for that just saved item. I just want to make sure that we are comfortable that this behavior will be evident to users.

**Comment by sdash-redhat** [comment]
Thanks Todd. This is not overthinking rather is positively helpful :).

I think the answer can be having "**Save**" and  "**Save and Open**" it is in the same line as what you are expecting and the users will be clear that the content is saved and then opened. It is exactly similar lines of "Add" and "Add & Open" but the saving will make it evident that the action makes the data to be actually saved. 

![image](https://user-images.githubusercontent.com/27761023/32318958-e8b29222-bfde-11e7-8b5c-7a658aeb8dcd.png)

Now when the user lands on the details page, he knows that the WI is created and saved, so any changes he makes cancel button will discard that and will close .

**Comment by joshuawilson** [comment]
I know we are not "mobile-first" yet, but I still need to ask. How would this work on a mobile devise?

**Comment by sdash-redhat** [comment]
This current design does not reflect the mobile viewport based rendering and nor even "mobile-first". That will be detailed out when the mobile-first is in the scope of delivery.

**Comment by joshuawilson** [comment]
At Java One, 90% of the user signed up for OSIO on their phones. This only worked because @mindreeper2420 used mobile-first for the login page. I'm afraid that if we follow this approach we will have mobile-last and this might make it hard to introduce it later when it will become part of the scope.

**Comment by sdash-redhat** [comment]
@joshuawilson is ur question for this specific deisgn feature discussed in this thread or the whole planner UX aspect?

Also, are you referring to "small screens" as mobility devices?

**Comment by joshuawilson** [comment]
I just wanted to bring it up again. I have been asking for mobile-first for 1.5 yrs on this project. I guess it is still out of scope.

**Comment by sdash-redhat** [comment]
@joshuawilson +1
I do agree with you that "mobile-first" is an important aspect. 
I am positive and do hope that in coming trains this would be addressed.

**Comment by mindreeper2420** [comment]
My 2 cents - For those thinking "mobile later, or we'll do mobile in the future", that is not as easy at it sounds. If a system is not designed or coded with mobile as even a thought from the start, going back will be next to impossible. Almost all of the Platform UI works on mobile devices (not perfect, but is usable), which was a conscience choice. Even the navigation has been retouched in order to work on mobile devices. The marketing page was designed and coded with mobile first, as that is the primary entry into OpenShift.io for those who are introduced to it at conferences.

We've never been "no mobile" or "forget about mobile", but has always been "we need to keep mobile in mind". I know that some aspects are not feasible on mobile (or smaller screen devices), but it should never be left out of the overall thought. Bootstrap (which is out base) is mobile first and PatternFly (our design architecture) is built off of Bootstrap, so mobile is built into the system already.

**Comment by nimishamukherjee** [comment]
@openshiftio/uxd-team can you evaluate this story and let us know if we need to work on it.

**Comment by nimishamukherjee** [comment]
@openshiftio/uxd-team can we re-define the story scope?
We now have quick add on the backlog page, the board view, and the query page.
Also we could consider what actions a user would perform after creating a work item.

**Comment by serenamarie125** [comment]
Based on a conversation held on 2019-Jan-04 including Dev, PM and UX, the UX team will not be providing support for planner items. Removing the team/ux label and un-assigning UX folks. If priorities are changed in the future, please add the team/ux label
